---
title: "Data Visualisation R Project"
output: 
  html_document:
    toc: true
    number_sections: true
    code_folding: hide
    toc_float:
      collapsed: true
    highlight: textmate
---

```{=html}
<style type="text/css">

#TOC{
    float: right;
    width: 20%;
    margin-right: 20px;
    padding: 5px;
    border: 1px solid #ddd;
    background-color: #f9f9f9;
}

.r {
    font-family: 'Courier New', monospace; 
    font-size: 14px;           
}


 body{
  font-size: 12pt;
  font-family: Arial, sans-serif;
  line-height: 2.0;
}
 
  .header-info {
    display: flex;
    justify-content: space-between;
    margin-bottom: 20px;
}

.header-info .left, .header-info .center {
    flex: 1;
}

.header-info .center {
    text-align: center;
}

.left {
    font-size: 12px;
    margin-bottom: 30px;
    color: grey;
}

.left p {
    margin-top: 0px;
    font-size: 15px;
    color: black;
}

  .header-info h3 {
  margin: 0;
  font-size: 16px;
  color: grey;
}

  .header-info p {
  margin: 0;
  font-size: 16px;
  color: black;
}
</style>

<div class="header-info">
  <div class="left">
    <h3>AUTHOR</h3>
    <p>Damilola Akindayo</p>
  </div>
  <div class="center">
    <h3>AFFILIATION</h3>
    <p>Deggendorf Institute of Technology</p>
  </div>
</div>

<div class="left">
  <h3 style="font-size:150%;">PUBLISHED</h3>
  <p>July 14, 2024</p>
</div>
```
# General information

1.  To enhance the readability of the plots, the `code_folding` option is set to "hide" in the YAML header. This setting can be adjusted to "show" all codes at the very top of this document, or to show individual code chunks as needed throughout the project.

2.  For clarity and organization, the function or task performed by each code chunk is specified at the beginning of the code under different subheaders. This aims to minimize clutter from individual comments.

3.  The directory containing the project dataset is permanently set for use throughout the R Markdown document, ensuring consistent access to the data.

```{r setup, echo = TRUE, message = FALSE}
knitr::opts_knit$set(root.dir = "~/Downloads/DataVis_R_2024_FinalProject/")
```

# Libraries

Necessary packages for the project, which have been previously installed using `install.packages`, are loaded first. These packages include:

-   **Tidyverse:** For data tidying, manipulation, and wrangling, and it also supports visualization.
-   **ggplot2:** Useful for visualizing the data using various graph and plot methods.
-   **ggpubr:** Assists in formatting plots generated by the `ggplot2` package. One of its functions, `stat_cor`, is used to add the Pearson correlation coefficient to scatter plots.
-   **ggrepel:** Provides a clean method for geom labeling and helps to repel overlapping text labels in plots, such as in Volcano plots.
-   **patchwork:** An extension of `ggplot2` that is useful for combining multiple plots into a single layout.

------------------------------------------------------------------------

```{r echo =TRUE, message = FALSE}
#Loading libraries
library(tidyverse)
library(ggplot2)
library(ggpubr)
library(ggrepel)
library(patchwork)
```

# TCGA expression data

The Cancer Genome Atlas (TCGA) is an extensive initiative focused on cataloging genetic mutations linked to cancer through genome sequencing and bioinformatics. It has gathered and analyzed tumor and normal tissue samples from thousands of patients, covering over 30 types of cancer.

## Loading and preparing data

The dataset files (“tcga_cancer.csv” and “tcga_cancer_labels.csv”) for the exploratory data analysis were loaded directly from the current working directory.

```{r class.source='r', message = FALSE}
tcga_data <- read_csv("tcga_cancer.csv")
```

The first five rows of the loaded tcga dataset

```{r}
#To view the first five rows of the loaded data
head(tcga_data, 5)
```

------------------------------------------------------------------------

**Tasks Performed by the Code in This Session**

1.  Selecting Columns and Rows: The code selects all columns and the first 200 rows from the `tcga` dataset

```{r message=FALSE, class.source='r'}
Gene_subset <- tcga_data[1:200,]
```

2.  Rounding Columns to Two Decimal Digits: The columns `gene_1` and `gene_12` are rounded to 2 decimal digits to standardize the data format.

```{r class.source='r', message = FALSE}
Gene_subset$gene_1 <- round(Gene_subset$gene_1, digits = 2)
Gene_subset$gene_12 <- round(Gene_subset$gene_12, digits = 2)
```

3.  The first ten rows of the rounded 200-row dataset are viewed to ensure that the rounding and selection processes have been correctly implemented.

```{r class.source='r', message=FALSE}
Gene_subset %>%
  select(gene_1, gene_12)
```

**Tasks Performed by the Code in This Session**

1.  Loading/Reading the TCGA Label File: The code loads or reads the TCGA label file into the R environment, preparing the data for further analysis.

```{r class.source='r', message = FALSE}
tcga_labels <- read.csv("tcga_cancer_labels.csv", header = TRUE)
#header is set to *TRUE* to ensure that the original column names are kept as it is
```

2.  Selecting Specific Columns and Rows: The code selects all columns and the first 200 rows from the TCGA label dataset for exploratory analysis and subsequent processing.

```{r class.source='r', message = FALSE}
LabelSubset <- tcga_labels[1:200, ]

```

3.  Glimpse/overview of the selected 200 label dataset

```{r}
glimpse(LabelSubset)
```

## Plotting

### Expression distribution Gene1

A histogram divides the x-axis into equally spaced bins and uses the height of a bar to display the number of observations that fall in each bin.

1.  **First Histogram Plot:** The bin width is set to "0.4" to match the provided sample plot.

------------------------------------------------------------------------

```{r message=FALSE, class.source='r'}

gene1_plot1 <- ggplot(Gene_subset, aes(x=gene_1)) + 
  geom_histogram(binwidth = 0.4, color = "black", fill="grey80",alpha = 0.6) +
  coord_cartesian(expand = FALSE, ylim = c(0,45)) +  #The expand function is used for the plots to ensure it extends way to the plot axes and border line
  
  #Plot labels and theme setting
  labs(title = "Expression Gene 1", x = "RNA counts", y = "Count") +
  theme_minimal() +  #theme_minimal is used for better default axis lines
  theme(
  plot.title = element_text(size = 14, lineheight = 3.5, colour = "black", 
  hjust = 0, vjust = 1.8), 
  axis.title.x = element_text(size = 14, lineheight = 1, face = "bold", 
  colour = "black"),
  axis.title.y = element_text(size = 14, lineheight = 1, face = "bold", 
  colour = "black", angle = 90, hjust = 0.5),
  axis.ticks = element_line(linewidth = 0.8, color = "black"),
  axis.ticks.length = unit(0.1, "cm"),
  axis.text = element_text(size = 12),
  axis.title = element_text(size = 15, face = "bold"),
  legend.position = "top",
  legend.ticks = element_blank(),
  panel.grid = element_blank(),
  panel.border = element_rect(linewidth = 1.8, fill = NA))
  
```

```{r gene1_plot1, fig.width = 3.8, fig.height = 3.8, message = FALSE}
plot(gene1_plot1)
```

2.  **Second Histogram Plot:** The bin width is set to "0.2" for a finer resolution of the data distribution.

```{r class.source='r', message = FALSE}
gene1_plot2 <- ggplot(Gene_subset, aes(x=gene_1)) + 
  geom_histogram(binwidth = 0.2, color = "black", fill="grey80",alpha = 0.6) +
  coord_cartesian(expand = FALSE , ylim = c(0,45)) +  #The expand function is used for the plots to ensure it extends way to the plot axes and border line
  
  #Plot labels and theme setting
  labs(title = "Expression Gene 1", x = "RNA counts", y = "Count") +
  theme_minimal() +  #theme_minimal is used for better default axis lines
  theme(
  plot.title = element_text(size = 14, lineheight = 3.5, colour = "black", 
  hjust = 0, vjust = 1.8), 
  axis.title.x = element_text(size = 14, lineheight = 1, face = "bold",
  colour = "black"),
  axis.title.y = element_text(size = 14, lineheight = 1, face = "bold", 
  colour = "black", angle = 90, hjust = 0.5),
  axis.ticks = element_line(linewidth = 0.8, color = "black"),
  axis.ticks.length = unit(0.1, "cm"),
  axis.text = element_text(size = 12),
  axis.title = element_text(size = 15, face = "bold"),
  legend.position = "top",
  legend.ticks = element_blank(),
  panel.grid = element_blank(),
  panel.border = element_rect(linewidth = 1.8, fill = NA))
```

```{r gene1_plot2, fig.width = 3.8, fig.height = 3.8, message = FALSE}
plot(gene1_plot2)
```

**Histogram Comparison**

-   **Histogram with Binwidth = 0.4:** This histogram offers a better balance for identifying patterns and interpreting the distribution of counts for the gene. The central peak around a count of 4 is clearly visible, making it easier to understand the overall distribution.

-   **Histogram with Binwidth = 0.2:** Although this histogram provides a more detailed view, it results in too many bars. The increased number of bars makes it difficult to visualize the overall shape of the distribution, potentially obscuring key patterns.

### Gene expression correlation

**Tasks Performed by the Code in This Session**

1.  Scatterplot to Visualize Relationship: A scatterplot is created to visualize the relationship between Gene 1 and Gene 12 of the dataset.

```{r class.source='r', message = FALSE}
Gene1_Gene12 <- ggplot(Gene_subset, aes(x=gene_1, y=gene_12)) + geom_point(shape = 21, size=3.0,color="black", fill ="lightblue", alpha = 0.9)
```

2.  Adding Linear Regression Line and Pearson Correlation Coefficient: The `geom_smooth` function is used to add a linear regression line to the scatterplot, while the `ggpubr` package is utilized to include the Pearson correlation coefficient on the scatterplot.

```{r class.source='r', message = FALSE}
gene1_gene12_plot <- Gene1_Gene12 + geom_smooth(method = "lm", se = TRUE, color = "black", linetype = 2) +
  stat_cor(method = "pearson", aes(label = after_stat(r.label)), label.x = 4.9, label.y = 4.55) + 
  #after_stat helps to ensure only the r value is displayed excluding the P-value
  
  #Defining labels and theme
  labs(title = "Relationship Gene 1 and Gene 12", x= "Gene 1", y ="Gene 12")+
  theme(
  plot.title = element_text(size = 14, hjust = 0, vjust = 1.0),
  axis.text = element_text(size = 12, colour = "black"),
  axis.title = element_text(size = 15, face = "bold"),
  axis.ticks = element_line(linewidth = 0.8, color = "black"),
  axis.ticks.length = unit(0.2, "cm"),
  panel.grid.minor.y = element_blank(),
  panel.grid.minor.x = element_blank(),
  panel.background = element_rect(fill = NA),
  panel.grid.major = element_line(colour = "grey90"),
  panel.border = element_rect(linewidth = 1.8, fill = NA))
```

```{r gene1_gene12_plot, fig.width = 4, fig.height = 4, message = FALSE}
plot(gene1_gene12_plot)
```

**1. Pearson correlation coefficient**

The Pearson correlation coefficient (R-value) measures the strength and direction of the linear relationship between two variables, ranging from -1 to 1. It quantifies the extent to which two variables, such as gene_1 and gene_12, tend to lie along a straight line when graphed.

R-value of 0 or close to 0 indicates no linear relationship between the variables. R-value of 1 indicates a perfect positive linear relationship. R-value of -1 indicates a perfect negative linear relationship. In this plot, the **R-value is -0.062**, indicating a very weak negative linear relationship between gene_1 and gene_12, as it is very close to 0.

**2. Gray line surrounding regression line**

The grey area surrounding the regression line represents the confidence interval and tells how confident and degree of uncertainty around the estimated regression line. Narrow gray areas indicates that there is high confidence in the estimated relationship. In other words, the true relationship/value between both genes expressions is likely close to the regression line while a wide grey area indicates lower confidence in the estimated relationship.

As shown in the plot, the grey line width reflects that there is uncertainty or no true relationship between Gene 1 and Gene 12.

### Class proportions

**Tasks Performed by the Code in This Session**

1.  Calculating Percentage / Relative Proportion of Gene Classes: The code calculates the percentage or relative proportion of each gene class in the dataset.

```{r class.source='r', message = FALSE}
label_proportion <- LabelSubset %>%
  group_by(Class) %>%
  summarise(Count = n()) %>% #Counts each gene class occurrence
  mutate(Percentage = Count / sum(Count) * 100)# This creates a new column called Percentage 

```

2.  Glimpse of the calculated relative proportion of the Gene classes including the new `Percentage` column

```{r}
glimpse(label_proportion)
```

3.  Visualizing Proportions with Stacked Bar Charts: Stacked bar charts are used to visualize the proportions of each gene class. `coord_flip` function is used to change the orientation of the stacked bar chart from horizontal to vertical.

```{r class.source='r', message = FALSE}
gene_class <- ggplot(label_proportion, aes(x = "", y = Percentage, fill = Class)) +
  #x= is set to "" to create a bar chart where all data is represented in a single stacked bars.
  
  geom_bar(stat = "identity", position = "stack", width = 1, color = "black") + #identity creates bars for each class with heights corresponding to their percentages.
  
  coord_flip(expand = FALSE, ylim = c(0,100)) + #flips the bar chart so that the bars are laid side by side
  
  scale_fill_manual(values = alpha(c("#F4857D", "#7CB8DD", "darkseagreen2","#C5A7ED", "#FAC26F"),0.6), limits = c("PRAD", "LUAD", "KIRC", "COAD", "BRCA")) #Limits is used to set the order of coloring the gene
```

4.  Adding Absolute Numbers: The `geom_text` function is utilized to add or insert the absolute numbers for each class inside the individual bars of the chart, providing clear numerical information along with the visual representation.

```{r class.source='r'}
geneclass_plot <- gene_class + 
  geom_text(aes(label = Count), position = position_stack(vjust = 0.5), size   = 6) + #to insert the values in the plot
  
  #Labels and theme setting
  labs(title = "Distribution of classes (n = 200)", y = "Percentage (%)", x =   NULL, fill = "Class") +
  theme_minimal() +
  theme(
    plot.title = element_text(size = 16),
    legend.title = element_text(size = 12),
    legend.text = element_text(size = 10),
    legend.position = "top",
    legend.direction = "horizontal",
    panel.grid = element_blank(),
    panel.border = element_rect(linewidth = 1.5, fill = NA),
    axis.title.x = element_text(size = 14, lineheight = .9, face = "bold",
    colour = "black"),
    axis.text = element_text(size = 13),
    axis.ticks.x = element_line(linetype = 2),
    axis.ticks.length = unit(0.2, "cm"),
    aspect.ratio = 0.12
  ) 
  
```

```{r geneclass_plot, fig.width = 7, fig.height = 4.5, message = FALSE, fig.show='asis'}
plot(geneclass_plot)
```

**Personal Opinion about the color and shade used**

The distinct colors used for each class (PRAD, LUAD, KIRC, COAD, BRCA) allow for easy differentiation at a glance, which helps to quickly understand the class distribution. However, the colors for PRAD (pink) and COAD (light purple) are quite similar, which could pose a problem for readers with color blindness. Adjusting or increasing the contrast, for example, by changing the alpha level from `0.6 to 0.9`, would be beneficial.

### Combining the plots

All three plots; Gene1 expression distribution, Gene1 and Gene 12 expression correlation, and Gene class proportions are combined into a single figure using `Patchwork` package

```{r class.source='r', message = FALSE}
combinedPlot1 <- gene1_plot1 + gene1_gene12_plot + geneclass_plot + plot_layout(design = "AABBBB\nAABBBB\nCCCCCC",widths = c(4,2), heights = c(8,16)) + plot_annotation(tag_levels = 'I') 
#plot_layout's design attribute helps to adjust the space taken by each plot in terms of rows and columns

```

```{r combindPlot1, fig.width = 7.5, fig.height = 7, echo=FALSE, message = FALSE}
plot(combinedPlot1)
```

# Human genome

GENCODE is a scientific initiative dedicated to providing a comprehensive and high-quality annotation of the human and mouse genomes. It helps to identify and classify all gene features, including protein-coding genes, non-coding RNAs, and pseudogenes.

## Loading and preparing data

```{r class.source='r', message = FALSE}
hg38_genes <- read.delim("hg38_genes.tsv", header = TRUE, sep = "\t", stringsAsFactors = FALSE)
```

An overview of the loaded human genome dataset content shows that it contains 7 columns and 20,000 rows

```{r class.source='r'}
glimpse(hg38_genes)
```

**Tasks Performed by the Code in This Session**

1.  Filtering Specific Gene Types: The code filters for genes whose type is either "protein_coding", "lncRNA", "processed_pseudogenes", or "transcribed_unprocessed_pseudogene".

```{r class.source='r', message = FALSE}
hg38_genes <- hg38_genes %>%
  filter(gene_type %in% c("protein_coding", "lncRNA", "processed_pseudogene", "transcribed_unprocessed_pseudogene"))
```

2.  Renaming and Combining Gene Groups: The "protein_coding" genes are renamed to "mRNA". The "processed_pseudogenes" and "transcribed_unprocessed_pseudogene" groups are combined to create a new group called "Pseudo".

```{r class.source='r', message = FALSE}
hg38_genes <- hg38_genes %>%
  mutate(gene_type = case_when(
    gene_type == "protein_coding" ~ "mRNA",
    gene_type %in% c("processed_pseudogene", 
                     "transcribed_unprocessed_pseudogene") ~ "Pseudo",TRUE ~ gene_type))
  
#TRUE ~gene_type is specified for any other values of gene_type not changed to retain its original value

```

3.  Filtering by Gene Length: The code keeps only genes with a length (distance between start and end) greater than 1,000 and less than or equal to 20,000.

```{r class.source='r', message = FALSE}
Subset_Genes <- hg38_genes %>%
  mutate(length = end_gene - start_gene) %>%
  filter(length > 1000 & length <= 20000)
```

4.  Overview of the new processed subset human genome dataset. This shows the dataset now contain 6,850 entries prior to 20,000 entries. It also includes the length column.

```{r class.source='r'}
glimpse(Subset_Genes)
```

## Plotting

### Gene count

**Tasks Performed by the Code in This Session**

1.  Summarizing the Total Count of Genes per Class: The code summarizes the total count of genes for each class in the dataset.

```{r class.source='r', message = FALSE}
class_gene <- Subset_Genes %>%
  group_by(gene_type) %>%
  summarise(Count = n())%>%
  arrange(Count) # arrange function is used so that the bar plot can plotted ranging from smallest to largest count

```

2.  The first ten rows of group gene classes and corresponding counts

```{r}
head(class_gene)
```

3.  Plotting a Classical Bar Chart: Using the newly formed categories, a classical bar chart is plotted. An appropriate range of the scale is used to accurately represent the data in the bar chart.

```{r class.source='r', message = FALSE}
gene_countPlot <- ggplot(class_gene, aes(x = reorder(gene_type, Count), y = Count, fill = gene_type)) + 
  geom_bar(stat = "identity", color = "black") + #fill = c("darkslateblue", "chocolate3", "aquamarine4"
  scale_fill_manual(values = alpha(c("#3FBC8B", "#D0740D", "#745EA7"), 0.8)) +
  scale_y_continuous(expand = c(0, 0), limits = c(0, 4000)) +  #Sets y-axis limits and expands plot to border
  
  #Labels and theme setting
  labs(title = "Total gene number", x = "Class", y = "Count") + 
  theme_bw() + 
  theme(
    panel.border = element_rect(linewidth = 0.5, fill = NA),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    panel.grid.minor.y = element_line(linetype = 0), 
    panel.grid.major.y = element_line(linetype = 2, color = "grey70", 
    linewidth  = 0.6),
    legend.position = "none",
    axis.text = element_text(size = 12, colour = "black"),
    axis.text.x = element_text(angle = 45, hjust = 1),
    plot.title = element_text(size = 16, lineheight = .9, colour = "black"),
    axis.title.x = element_text(size = 14, lineheight = .9, face = "bold", colour = "black"),
    axis.title.y = element_text(size = 14, lineheight = .9, face = "bold", colour = "black")
    )

```

```{r gene_countPlot, fig.width = 4, fig.height = 4, message = FALSE}
plot(gene_countPlot)
```

### Length distribution

**Tasks Performed by the Code in This Session**

1.  Creating an Overlapping Density Plot: An overlapping density plot is created to visualize how gene length (distance between start and end gene) is distributed in each class.

```{r class.source='r', message = FALSE}
#overlapping density plot to look at how the gene length is distributed in each class 
Length_Plot <- ggplot(Subset_Genes, aes(x = length, fill = gene_type)) +
  geom_density(alpha=0.5) +
  scale_fill_manual(values = c("#3FBC8B","#D0740D", "#745EA7")) +
  scale_y_continuous(expand = c(0, 0)) +  # Set y-axis limits and remove expansion
  guides(fill = guide_legend(title = NULL, reverse = TRUE))  #reverses the legend order from default to match specified one 
```

2.  Calculating Median Gene Length for Each Class: The median of gene length for each class is calculated to add reference lines to the density plot.

```{r class.source='r', message = FALSE}

medianLength <- Subset_Genes %>%
  group_by(gene_type) %>% #Grouping is necessary before summarizing to get the median
  summarize(median_Gene_type = median(length))

```

3.  Glimpse of the calculated median of different Gene type/class

```{r}
glimpse(medianLength)
```

4.  Adding Median Lines: The `geom_vline` function is used to add vertical lines representing the calculated medians to the prepared density plot.

```{r class.source='r', message = FALSE}
#To include median length in the plot using geom_vline.
gene_lengthPlot <- Length_Plot + geom_vline(data = medianLength, aes(xintercept = median_Gene_type), color = c("#8363C3","#36CA90", "#C74104"), linetype = c("solid","dashed", "dashed"), linewidth = c(0.8, 0.8, 0.5)) + 
  
  #Labels and theme definition
  labs(x="Gene Length (bp)", y = "Density", title = "Lenght distribution of gene classes" ) +
  theme_void() +
  theme(
    axis.title.x = element_text(size = 14, lineheight = .9, face = "bold",
    colour = "black"),
    axis.title.y = element_text(size = 14, lineheight = .9, face = "bold", 
    colour = "black", , angle = 90, hjust = 0.5),
    plot.title = element_text(size = 16, lineheight = .9, colour = 
    "black"),
    axis.text = element_text(size = 14),
    panel.grid.minor.y = element_blank(),
    panel.grid.minor.x = element_blank(),
    panel.grid.major = element_line(colour = "grey90"),
    panel.background = element_blank(),
    panel.border = element_blank(),
    legend.position = "inside", legend.justification = c(0.9, 1))

```

```{r gene_lengthPlot, fig.width = 6.5, fig.height = 4, message = FALSE}
plot(gene_lengthPlot)
```

5.  **Boxplot for Length Distribution Comparison:** A boxplot is to compare the length distribution of genes across different classes.

```{r class.source='r'}
ggplot(Subset_Genes, aes(x = factor(gene_type), y = length, fill = gene_type)) +
  geom_boxplot(outlier.size = 1.5, outlier.shape = 21, alpha = 0.7) +
  scale_fill_manual(values = c("#3FBC8B","#D0740D", "#745EA7")) +
  labs(x="Gene class", y = "Length", title = "Distribution of gene classes")+
  theme_minimal() +
  theme(
    axis.title.x = element_text(size = 14, lineheight = .9, face = "bold",
    colour = "black"),
    axis.title.y = element_text(size = 14, lineheight = .9, face = "bold", 
    colour = "black", , angle = 90, hjust = 0.5),
    plot.title = element_text(size = 16, lineheight = .9, colour = 
    "black"),
    axis.text = element_text(size = 14),
    panel.grid.minor.y = element_blank(),
    panel.grid.minor.x = element_blank(),
    panel.grid.major = element_line(colour = "grey90"))

```

**Comparing overlapping density plot and box plot**

The overlapping density plot helps to easily visualize and compare overall distribution of gene lengths across different classes. It also allows for easy identification of peaks and the spread of data. As shown in the plot, Pseudo genes have a sharp peak at shorter lengths, indicating a high frequency of short pseudo genes at this point while mRNA and lncRNA have broader distributions.

Personally, I find the boxplot more intuitive to read and understand. It automatically provides a clear and straightforward way to compare central tendencies (such as the median), variability (interquartile range (IQR), first quartile (Q1), and third quartile (Q3)), and outliers between different classes. The boxplot vividly shows that pseudogenes have several outliers with significantly longer lengths, indicating high variability in the length of pseudogenes.

This clear visualization of differences in gene length distributions with boxplot helps to identify areas for further investigation. For instance, the presence of many outliers in pseudogenes might prompt a decision to normalize the data or perform Differential Expression Analysis (DEA).

### Gene Count and Gene Length Figure

Combined plot of the gene count and gene length distribution.

```{r echo = TRUE, class.source='r', message = FALSE}

combinedPlot2 <- gene_countPlot + gene_lengthPlot + plot_annotation( tag_levels = 'A', caption = "Figure 1: Overview of human genes. A: Total number per gene class. B: Gene length distribution per class.\nVertical lines represent median gene length per class.",theme = theme(plot.caption = element_text(size = 8, hjust = 0, colour = "grey40"))
  ) + 
  plot_layout(design = "ABBB\nABBB")

```

```{r combindPlot2, fig.width = 8, fig.height = 5, echo=FALSE, message = FALSE}
plot(combinedPlot2)
```

# Differential gene expression

Differential gene expression (DGE) analysis is a method used to identify genes that show significant changes in expression levels across various conditions, such as different cell types, disease states, or experimental treatments. This analysis helps in understanding how gene expression is altered under different biological contexts.

## Loading and preparing data

Loading the dummy deseq dataset using "read.table"

```{r class.source='r', message = FALSE}
deseq2_dummy <- read.table("deseq2_dummy.txt", header = TRUE)
```

An overview of the individual variables of the DEseq2 output shows that the file contains 14 columns and 5,000 rows

```{r}
glimpse(deseq2_dummy)
```

**Tasks Performed by the Code in This Session**

1.  Selecting the Average of Normalized RNA-seq Count Values Row-wise: The code calculates the average of the normalized RNA-seq count values for the two conditions analyzed (3 replicates of normal and cancer each).

    -   The `starts_with` helper function is used to select all normalized RNA-seq counts starting with "cancer" or "normal".

    -   The `rowMeans` function is used to calculate the average of these values across each row.

```{r class.source='r', message = FALSE}
normalized_columns <- select(deseq2_dummy, starts_with("cancer_"), starts_with("normal_"))
deseq2_dummy$avg_normalized_count <- rowMeans(normalized_columns, na.rm = TRUE)
```

2.  Ordering the Data Frame: The data frame is ordered by the average normalized count in descending order to prioritize the highest average counts.

```{r class.source='r', message = FALSE}
ordered_Avg <- deseq2_dummy[order(deseq2_dummy$avg_normalized_count, decreasing = TRUE), ]
```

3.  Selecting the Top 2000 Genes: The top 2000 genes are selected based on the average of the normalized count values.

```{r class.source='r', message = FALSE}
top2000Genes <- ordered_Avg[1:2000, ]
```

## Plotting

### Volcano plots

The volcano plot visually represents differentially expressed genes, with the most highly up-regulated genes displayed on the right side, the most highly down-regulated genes on the left side, and the most statistically significant genes positioned at the top.

**Tasks Performed by the Code in This Session**

1.  Defining the thresholds using specified values

```{r class.source='r', message = FALSE}
top2000Genes$Significant_Genes <- "n.s."

top2000Genes$Significant_Genes[top2000Genes$padj < 0.1 & top2000Genes$log2FoldChange > 0] <- "up"

top2000Genes$Significant_Genes[top2000Genes$padj < 0.1 & top2000Genes$log2FoldChange < 0] <- "down"
```

2.  Adding Reference Lines:

    -   The `geom_hline` function is used to add a horizontal reference line indicating the significance threshold for the p-value.

    -   The `geom_vline` is used to add a vertical reference line indicating the significance threshold for the fold change

3.  Labeling Significant Genes :

    -   The `geom_label_repel` function from the`ggrepel`package is used to add labels and connecting lines to significant genes.

    -   Labels are added to genes with a log2 fold change higher than 2 and an adjusted p-value lower than 0.1, ensuring clear identification of these significant genes on the plot.

```{r class.source='r', message = FALSE}
deg_Plot <- ggplot(top2000Genes, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(fill = Significant_Genes), alpha = 0.8, shape = 21, color = "black", size = 2) + 
  scale_fill_manual(values = c('up' = "#850626", 'n.s.' = "white", 'down' = 'skyblue')) +
  scale_y_continuous(expand = c(0.01, 0)) +
  
  #geom_hline and geom_vline to add reference lines
  geom_hline(yintercept = -log10(0.1), linetype = "dotted") +
  geom_vline(xintercept = 0, linetype = "solid") +
  
  #ggrepel is used to add label to significant genes significance threshold
  geom_label_repel( 
    data = subset(top2000Genes, padj < 0.1 & abs(log2FoldChange) > 2), 
    aes(label = gene_id),
    size = 3.5,
    box.padding = 0.5,
    point.padding = 0.8,
    force = 10,
    nudge_y = 0.8, #adjusts length of connection line
    max.overlaps = Inf,
    min.segment.length = 0.05
  ) +
  
  #Plot labels and themes setting
  labs(title = "Diff. expressed genes (Cancer vs. Normal)", x = "Fold change (log2)", y = "Adjusted p-value (-log10)", fill = "Regulation") +
  theme_minimal() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(linewidth = 2, fill = NA),
    plot.title = element_text(size = 16),
    axis.text = element_text(size = 14, colour = "black"),
    axis.title = element_text(size = 14, face = "bold"),
    axis.ticks = element_line(linetype = 1),
    axis.ticks.length = unit(0.2, "cm"),
    legend.position = "inside",
    legend.position.inside = c(0.91, 0.91),
    legend.text = element_text(size = 13),
    legend.box.background = element_rect(colour = "black", fill = NA,
    linewidth = 0.8))
  

```

```{r deg_Plot, fig.width = 7, fig.height = 8, echo=FALSE, message = FALSE}
plot(deg_Plot)
```

### Heatmaps

Heatmap of the RNA-seq expression values for the 20 genes with the minimum adjusted p-value.

**Adjustments made to the code**

1.  Changing the Order By Parameter: The `order by` clause at the beginning of the code was changed from the string `"padj"` to the already available variable `padj` in the dataframe.

2.  Selecting Specific Columns: The "dplyr::select" function was used to select the `gene_id` column and 6 other columns (3 normal and 3 cancer), making it a total of 7 columns. As a result, setting "pivot_longer" to select columns `"9:14"` generated an error because there were only 7 columns being used. The correct columns `(2:7)` was specified to avoid errors.

3.  Adjusting scale_fill_distiller: The `scale_fill_distiller` function was altered to include the direction and breaks, enhancing the customization of the color scale in the plot.

4.  Additional changes made to the plot code are explained within the code using comments for clarity.

```{r}
deseq2_dummy
```

```{r echo = TRUE, include=TRUE, message = FALSE}
deseq2_res_f <- deseq2_dummy %>%
  slice_min(n = 20, order_by = padj)
#The order_by changed to padj instead of "padj".

heat_plot <- deseq2_res_f %>% 
  dplyr::select(gene_id, 9:14) %>%
  pivot_longer(cols = 2:7 , names_to = "condition", values_to = "counts") %>%
  #Adjusting for correct range to be selected; "2:7"
  
  ggplot(aes(x = condition, y = gene_id, fill = log2(counts), label = round(log2(counts), digits = 1))) + 
  #digit is changed to 1 decimal places instead of 2 to match with the specified one in the heatmap plot
  
  geom_tile(color = "black") +
  geom_text() +
  coord_cartesian(expand = FALSE) + #According to the error "the condition has length > 1". So expand was changed to False
  
  scale_fill_distiller(palette = "Purples", direction = 1, breaks = c(2.5, 7.5, 12.5)) + 
  # The direction of the legend is set to 1 making the most expressed genes, the darkest. The range of legend is also changed from 6 to contain just 3 characters using breaks
  
  labs(x = "Sample", y = "Gene IDs", fill = "Counts (log2)") +
  theme_minimal() +
  theme(
    legend.position = "top",
    legend.ticks = element_blank(),
    title = element_text(size = 13),
    axis.text = element_text(size = 11, colour = "black"),
    axis.title = element_text(size = 15, face = "bold"),
    panel.border = element_rect(linewidth = 3, fill = NA)) 
   #changed size to linewidth instead as indicated by the error message
```

```{r heat_plot, fig.width = 6, fig.height = 9, echo=FALSE, message = FALSE}
plot(heat_plot)
```

**REFERENCES**

1.  R for Data Science, 2nd Edition Hadley Wickham
2.  Biocore CRG. (n.d.). *Volcano plots*. Retrieved July 6, 2024, from <https://biocorecrg.github.io/CRG_RIntroduction/volcano-plots.html>
3.  Wilke, C. (n.d.). *Patchwork layout guides*. Retrieved July 10, 2024, from <https://patchwork.data-imaginist.com/articles/guides/layout.html>
